ledger-ondisk/src/LedgerOnDisk/Diff.hs

```diff
@@ -73,13 +116,34 @@ applyDtoHashMap :: (Semigroup v, Eq k, Hashable k) => HashMap k (Diff v) -> Hash
 applyDtoHashMap d = appEndo (HashMap.foldMapWithKey go d) where
   go k = Endo . applyDforK k
 
+
+{- We probably want to use a strict left fold, to avoid creating thunks. Maybe the use of Endo is overkill?
 applyDifftoMap :: Semigroup v => Ord k => Map k (Diff v) -> Map k v -> Map k v
 applyDifftoMap d = appEndo (Map.foldMapWithKey go d) where
   go k = Endo . \case
     DChangeTo v -> Map.insert k v
     DMappend v -> Map.insertWith (<>) k v
     DRemove -> Map.delete k
     DNoChange -> id
+-}    
+
+applyDifftoMap :: (Semigroup v, Ord k) => Map k (Diff v) -> Map k v -> Map k v
+applyDifftoMap d m =  Map.foldlWithKey' accum m messages
+  where
+    accum ans key DRemove = Map.delete key ans
+    accum ans key (DChangeTo v) = Map.insert key v ans  -- No need to use (insertWith (<>)) we want to overide whatever is there
+    accum ans key (DMappend e) = Map.update (Just . (e <>)) key ans
+    accum ans key DNoChange = ans
+
+-- So if (Diff v) represent changesto 'v'
+-- then (Map k (Diff v)) represents changes to (Map k v)
+-- Just like we want to compose changes to 'v', we want to compose changes to (Map k v)
+-- We have (<>) :: Diff v -> Diff v -> Diff v
+-- So we want (<>) :: Map k (Diff v) -> Map k (Diff v) -> Map k (Diff v)
+-- We would like to write this
+-- instance (Ord k,Semigroup v) => Semigroup (Map k (Diff v)) where
+--    (<>) x y  = (Map.unionWith <> x y)
+-- But this is already instance, Because (Diff v) has a Semigroup instance!
```

Comment:
I would guess that using `Data.Strict.Map.foldMapWithKey` and `Endo` is roughly equivalent to `Data.Strict.Map.foldlWithKey'`, if you prefer the latter that's good enough for me.

You've also implemented the change to `DRemove <> DMappend _` here.

Not sure if I'm misunderstanding you, but the `Semigroup` instance for `Map k v` does not use `v`'s `Semigroup` instance. It uses `union`, which uses the left value if both values are present.

If you look at `DiffMap` you can see it's `Semigroup` instance uses `MonoidalMap` (from https://hackage.haskell.org/package/monoidal-containers).

Reply:
Actually, for some reason it isn't :/ but, the Semigroup instance for Diffmap should (IMO) be written
```
deriving via (MonoidalMap k (Diff v)) instance Ord k => Semigroup (DiffMap k v)
```

Reply:
What I am complaining about is that we get overlapping instances. My instance is the one I want, the one supplied by haskell does not have the prpperties I want:   apply (d1 <> d2) m) to be apply d1 (apply d2 m)

Reply:
Yeah it's an annoying aspect of `containers`. We need to use a newtype for the Map for this reason
```haskell
newtype DiffMap k v = DiffMap (Map k (Diff v))
  deriving stock (Show, Eq)
  deriving (Semigroup, Monoid) via MonoidalMap k (Diff v)
```